### ISSUE TITLE
add storage server state handling for single clusters

### ISSUE THREAD

**Issue by GlenDC** [zero-os/0-Disk#603]
Currently we simply overwrite configs for single clusters (Primary/Template/Slave), which aren't used as pairs or in combination with other clusters. This in locations where we require HotReloading, state handling and anything related (for now just the NBDserve,r and just the Primary- and Template clusters).

+ [x] We do not want to ignore config updates, even the invalid ones, as it cannot be assumed that the old configs are still working;
+ ~When an online server gets replaced by another online server (in case of primary and slave cluster), we should copy the vdisk data from the old to the new one, as there might otherwise be a window where the new one might be missing data (NBDServer takes full responsibility here, until something goes wrong) (for template cluster this feature is not supported, as we use template cluster as a read-only cluster);~
+ [x] State/DialConfig changes of existing server (indices) should be handled individually to ensure, we can inform AYS about anything important, to handle any thing we have to handle, and to prevents things that shouldn't be happened when possible (e.g. a server marked as `RIP` should never change state again, as this is not supported, we can obviously only guarantee such things on a 0-Disk level for vdisks which are mounted and active);
+ ~AYS should be properly informed about any storage server updates (informational or error-wise because we are now in an invalid state);~
+ We do not currently provide the possibility to shrink/grow a cluster of an active vdisk (as in one that is currently mounted in the nbdserver), so this should not be allowed (and AYS should be notified that this happened);
   + [x] growing can be ignored without any serious consequences;
   + [x] when shrinking however, we'll mark all "deleted" servers as state unknown. This might in turn cause a vdisk read/write failure in case the vdisk touches these deleted servers (which is 99,999% of the times going to happen if they are active);
       + (we do this, because we assume that AYS is in control, and thus we shouldn't try to outsmart it by using an out-of-date config that might no longer be valid, after all there is probably a good reason why the config _had_ to be updated);

**Comment by GlenDC** [comment]
A lot of this work was already done as part of my work on the primary-slave-cluster-pair feature in the https://github.com/zero-os/0-Disk/tree/nbd-cluster-pair branch.

The only open tasks left to do for this issue are blocked by open issues #595 and #601.

**Comment by GlenDC** [comment]
Just to clarify why growing and shrinking of clusters (in terms of server size) is not going to be supported for active (see: mounted) vdisks...

For shrinking you could say, hey why not just re-spread all data from the now-to-be-deleted servers? Well that wouldn't be possible as we can't assume that the server configs that were deleted are still alive, remember: AYS is always right. So you could say, well than, why not first force AYS to mark all the servers that will be trimmed of as state `respread`, such that the nbdserver's vdisk storage logic can respread the data from those vdisks and than mark them as `rip`. Once AYS is notified about all `RIP` servers it wants it could safely shrink the cluster, right? Nope! Because one would still require to redistribute all data from the existing servers, as the spreading algorithm results will be different now because the difference in server size. This is an operation not yet supported, and on top of that I think we shouldn't do this heavy-lifting work for an active vdisk.

It's because the missing redistribution of data feature, that we also do not support growing of clusters. That and the fact that I think we shouldn't do this heavy-lifting work for an active vdisk.

**Comment by GlenDC** [comment]
At some point we should probably also add this to the docs, which I should be reminded of to do so in the near future when all this code lives on master.

**Comment by GlenDC** [comment]
> When an online server gets replaced by another online server (in case of primary and slave cluster), we should copy the vdisk data from the old to the new one, as there might otherwise be a window where the new one might be missing data (NBDServer takes full responsibility here, until something goes wrong) (for template cluster this feature is not supported, as we use template cluster as a read-only cluster);

This is for now cancelled, due to the missing feature of being able to do a DeepCopy of a vdisk for a cluster/server. Also issue #612 should be answered before we can focus on this state change.

> AYS should be properly informed about any storage server updates (informational or error-wise because we are now in an invalid state);

As we pause work for now on 0-Disk, we'll keep this feature as part of the other issue #595, as it is strongly related to that issue (in fact this is just an implementation of what the issue subscribes, so this is in fact the same issue if you will).

**Comment by GlenDC** [comment]
For now we'll make use of the single-cluster work done in the nbd-cluster-pair branch, so it doesn't get lost. Once we get back to our 0-Disk work however we should resolve #615 ASAP. As it will be required for the missing parts of those clusters, as well as well as the pair clusters still to be implemented (e.g. primary-slave clusters).
